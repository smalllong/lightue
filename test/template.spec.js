var L = require('../dist/cjs'),
  { div, span, section, input } = L

describe('template', () => {
  it('create root element', () => {
    L(div.root('hello', ' world'))
    expect(document.body).toMatchSnapshot()
  })

  it('create child elements', () => {
    L(div.hi('hello'), div.foo(div.barBaz('123')), div(345), section(), 'yes')
    expect(document.body).toMatchSnapshot()
  })

  it('not create null or undefined', () => {
    L(div.root(null, undefined))
    expect(document.body).toMatchSnapshot()
  })

  it('not show $if', () => {
    L(
      div.foo({ $if: false }, 'bar'),
      div.undefined({ $if: undefined }, 'should not be in DOM'),
      div.aaa({ $if: true }, 'bbb')
    )
    expect(document.body).toMatchSnapshot()
  })

  it('create child span element', () => {
    L(span.hi('hello'), div.foo(span.bar(234)))
    expect(document.body).toMatchSnapshot()
  })

  it('simple array', () => {
    L(div.root(...[1, 2, 3, 4].map((n) => div(n))))
    expect(document.body).toMatchSnapshot()
  })

  it('$class', () => {
    L(
      div.aaa(
        {
          $class: { foo: 1, bar: 0 },
        },
        'bbb'
      )
    )
    expect(document.body).toMatchSnapshot()
  })

  it('$value', () => {
    var vm = L(input.aaa({ $value: 'hello' }))
    expect(vm.el.children[0].value).toBe('hello')
  })

  it('$checked', () => {
    var vm = L(input.aaa({ type: 'checkbox', $checked: 1 }))
    expect(vm.el.children[0].checked).toBe(true)
  })

  it('attributes', () => {
    L(
      input.aaa({
        type: 'checkbox',
        dataFoo: 'bar',
      })
    )
    expect(document.body).toMatchSnapshot()
  })

  it('listener', () => {
    var onclick = jest.fn()
    var vm = L(
      div.root({
        onclick,
      })
    )
    vm.el.children[0].dispatchEvent(new MouseEvent('click'))
    expect(onclick).toHaveBeenCalled()
  })

  it('$cleanup', () => {
    var cleanup = jest.fn(),
      S = L.useState({
        curComp: 0,
      })
    function comp1() {
      return div({
        $cleanup: cleanup,
      })
    }
    function comp2() {
      return div()
    }
    var comps = [comp1, comp2]
    L(() => comps[S.curComp]())
    setTimeout(() => {
      expect(cleanup).not.toHaveBeenCalled()
      S.curComp = 1
      setTimeout(() => expect(cleanup).toHaveBeenCalled())
    })
  })
})
