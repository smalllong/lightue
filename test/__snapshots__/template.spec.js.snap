// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template $class 1`] = `
<body>
  <div
    class="aaa foo"
  >
    bbb
  </div>
</body>
`;

exports[`template attributes 1`] = `
<body>
  <input
    class="aaa"
    data-foo="bar"
    type="checkbox"
  />
</body>
`;

exports[`template create child elements 1`] = `
<body>
  <div
    class="hi"
  >
    hello
  </div>
  <div
    class="foo"
  >
    <div
      class="barBaz"
    >
      123
    </div>
  </div>
  <div>
    345
  </div>
  <section />
  yes
</body>
`;

exports[`template create child span element 1`] = `
<body>
  <span
    class="hi"
  >
    hello
  </span>
  <div
    class="foo"
  >
    <span
      class="bar"
    >
      234
    </span>
  </div>
</body>
`;

exports[`template create root element 1`] = `
<body>
  <div
    class="root"
  >
    hello
     world
  </div>
</body>
`;

exports[`template multiple array 1`] = `
<body>
  <div
    class="root"
  >
    <div>
      1
    </div>
    <div>
      2
    </div>
    <div>
      3
    </div>
    <div>
      4
    </div>
    <div>
      5
    </div>
    <div>
      6
    </div>
    <div>
      7
    </div>
  </div>
</body>
`;

exports[`template not create null or undefined 1`] = `
<body>
  <div
    class="root"
  />
</body>
`;

exports[`template not show $if 1`] = `
<body>
  <!---->
  <!---->
  <div
    class="aaa"
  >
    bbb
  </div>
</body>
`;

exports[`template simple array 1`] = `
<body>
  <div
    class="root"
  >
    <div>
      1
    </div>
    <div>
      2
    </div>
    <div>
      3
    </div>
    <div>
      4
    </div>
  </div>
</body>
`;
